### PR TITLE
Assemble Linear QFunction

### PR DESCRIPTION
**Pull request by jeremylt**
~This WIP PR adds assembly of the QFunction.

ToDo:
- [x] QFunction in ref
- [x] QFunction diagonal in ref
- [x] Nonlinear QFs
- [x] Refactor into operator code
~- [ ] Dispatch to ref~
- [x] Tests

This is an intermediate step in supporting preconditioners.~
This will be reworked following the efforts in #304. Also, I will focus on assembling linear QFunctions for now to support next month's talks.

### CODE REVIEW

**Comment by jeremylt** [comment]
This works for blocked and serial ref. Unfortunately, dispatching to ref is a bit messier than I anticipated.

File: backends/ref/ceed-ref-operator.c [hunk 1]
```diff
@@ -239,6 +222,179 @@ static int CeedOperatorApply_Ref(CeedOperator op, CeedVector invec,
       CeedChk(ierr);
     }
   }
+  return 0;
+}
+
+// Input basis action
+static inline int CeedOperatorInputBasis_Ref(CeedInt e, CeedInt Q,
```
Comment (jeremylt) [inline backends/ref/ceed-ref-operator.c]: I did some substantial rearrangement, breaking the operator code into chunks for reuse.

File: backends/ref/ceed-ref-operator.c [hunk 2]
```diff
@@ -435,6 +496,197 @@ static int CeedCompositeOperatorApply_Ref(CeedOperator op, CeedVector invec,
...
+
+    // Linearize QFunction
+    ierr = CeedQFunctionApply(qf, Q, impl->qvecsin, impl->qvecsout);
+    CeedChk(ierr);
+    // Copy A
+    for (CeedInt out=0; out<numactiveout; out++)
+      for (CeedInt q=0; q<Q; q++)
+        a[(e*numactiveout+out)*Q+q] = activeout[out][q];
+    // Compute B
+    CeedScalar epsilon = 1E-6, epsinv = 1E6;
```
Comment (jeremylt) [inline backends/ref/ceed-ref-operator.c]: This is *a* finite differencing to linearize, but I'm not entirely happy with the loss of accuracy. I should look for a better way to decide upon epsilon.

File: include/ceed.h [hunk 3]
```diff
@@ -349,6 +349,8 @@ CEED_EXTERN int CeedOperatorSetField(CeedOperator op, const char *fieldname,
                                      CeedVector v);
 CEED_EXTERN int CeedCompositeOperatorAddSub(CeedOperator compositeop,
     CeedOperator subop);
+CEED_EXTERN int CeedOperatorLinearizeQFunction(CeedOperator op,
```
Comment (jeremylt) [inline include/ceed.h]: I could add an interface to handle qfunctions that are already linear. I wanted to start with something more general.

File: tests/t511-operator.c [hunk 4]
```diff
@@ -1,6 +1,6 @@
 /// @file
-/// Test creation creation, action, and destruction for mass matrix operator
-/// \test Test creation creation, action, and destruction for mass matrix operator
+/// Test creation, action, and destruction for mass matrix operator
```
Comment (jeremylt) [inline tests/t511-operator.c]: Spelling is hard.

File: tests/t530-operator.c [hunk 5]
```diff
@@ -0,0 +1,191 @@
+/// @file
```
Comment (jeremylt) [inline tests/t530-operator.c]: This test checks the linearization of a linear qfunction. I test both the point values and the action of the linearization in the mass operator.

File: tests/t531-operator.c [hunk 6]
```diff
@@ -0,0 +1,197 @@
+/// @file
```
Comment (jeremylt) [inline tests/t531-operator.c]: This example tests a multi-field, nonlinear qfunction.

**Comment by jeremylt** [comment]
@jedbrown, I'd like to put in a 2D Poisson test, but otherwise this should be good (barring any bugs I might find in the test).

File: interface/ceed-operator.c [hunk 7]
```diff
@@ -210,6 +210,51 @@ int CeedCompositeOperatorAddSub(CeedOperator compositeop,
...
+/**
+  @brief Assemble a linear CeedQFunction associated with a CeedOperator
+
+  This returns a CeedVector containing a matrix at each quadrature point
+    providing the action of the CeedQFunction associated with the CeedOperator.
+    The vector 'assembled' is of size
+                                (size of output fields)*(size of input fields)*
+                                (num of quadrature pts)*(num of elements)
+    and contains row-major matrices representing the action of the
+    CeedQFunction for a corresponding quadrature point on an element.
```
[the remark this answers is not in the document]
Comment (jeremylt) [reply interface/ceed-operator.c]: I updated the language. I think this is clearer, and I put in a test, t532, that shows using multiple fields.

**Comment by jeremylt** [comment]
I will fix merge conflicts tomorrow.

**Comment by jeremylt** [comment]
Updated to master.

File: backends/ref/ceed-ref-operator.c [hunk 8]
```diff
@@ -434,6 +534,146 @@ static int CeedCompositeOperatorApply_Ref(CeedOperator op, CeedVector invec,
...
+    if (vec == CEED_VECTOR_ACTIVE) {
+      ierr = CeedQFunctionFieldGetSize(qfinputfields[i], &size); CeedChk(ierr);
+      ierr = CeedVectorGetArray(impl->qvecsin[i], CEED_MEM_HOST, &tmp);
+      CeedChk(ierr);
+      ierr = CeedRealloc(numactivein + size, &activein); CeedChk(ierr);
+      for (CeedInt field=0; field<size; field++) {
+        activein[numactivein+field] = &tmp[field*Q];
+      }
+      numactivein += size;
+      ierr = CeedVectorRestoreArray(impl->qvecsin[i], &tmp); CeedChk(ierr);
```
Comment (jedbrown) [inline backends/ref/ceed-ref-operator.c]: How do we know these arrays have been zeroed before we start using them below?

Also, I'm not wild about the action-at-a-distance of restoring the array. Probably better to do that down below, after using it.

File: tests/t530-operator.h [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+  const CeedScalar *weight = in[0], *J = in[1];
+  CeedScalar *rho = out[0];
+  for (CeedInt i=0; i<Q; i++) {
+    rho[i] = weight[i] * (J[i+Q*0]*J[i+Q*3] - J[i+Q*1]*J[i+Q*2]);
+  }
+  return 0;
+}
+
+CEED_QFUNCTION(mass)(void *ctx, const CeedInt Q, const CeedScalar *const *in,
+                     CeedScalar *const *out) {
```
[the remark this answers is not in the document]
Comment (jeremylt) [reply tests/t530-operator.h]: With OCCA not yet supporting the gallery, I've held off on converting our test suite to use the gallery. I'm planning on a little PR to change over our tests after #305.

File: backends/ref/ceed-ref-operator.c (hunk 8 above)
Comment (jeremylt) [reply backends/ref/ceed-ref-operator.c]: Added an explicit zero. I was just getting lucky by calling Assemble before OperatorApply in the tests.

We can't apply the QFunction without restoring access to these qvecs. I saw options as
1) Restore the array but keep a pointer (what I'm doing now)
2) Keep track of the vec and offset from the head of the array for each input field and use GetArray/RestoreArray on each loop

1 felt more straightforward and cleaner to me, but I'd be happy to switch to 2 if not bypassing the access system we set up is important.

File: tests/t532-operator.h [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+  // out[1] is output to multiply against gradient v, shape [2, nc=1, Q]
+  CeedScalar *v = out[0], *dv = out[1];
+
+  // Quadrature point loop
+  for (CeedInt i=0; i<Q; i++) {
+    const CeedScalar du0 = du[i+Q*0];
+    const CeedScalar du1 = du[i+Q*1];
+    v[i+Q*0] = qd[i+Q*0]*du0 + qd[i+Q*3]*du1 + qd[i+Q*6]*u[i];
+    dv[i+Q*0] = qd[i+Q*1]*du0 + qd[i+Q*4]*du1 + qd[i+Q*7]*u[i];
+    dv[i+Q*1] = qd[i+Q*2]*du0 + qd[i+Q*5]*du1 + qd[i+Q*8]*u[i];
```
[the remark this answers is not in the document]
Comment (jeremylt) [reply tests/t532-operator.h]: Ah, I put this in on purpose to demonstrate that rows/columns appear as given by the user.
In t532, `du` and `u` are inputs, in that order, and `v` and `dv` are added in that order. That's what I meant by `Inputs and outputs are in the order provided by the user when adding CeedOperator fields.`
I'll expand to be more clear.

File: interface/ceed-operator.c [hunk 11]
```diff
@@ -243,6 +243,52 @@ int CeedCompositeOperatorAddSub(CeedOperator compositeop,
   return 0;
 }
 
+/**
+  @brief Assemble a linear CeedQFunction associated with a CeedOperator
+
+  This returns a CeedVector containing a matrix at each quadrature point
+    providing the action of the CeedQFunction associated with the CeedOperator.
+    The vector 'assembled' is of shape
+      [num_elements, num_input_fields, num_output_fields, num_quad_points]
```
[the remark this answers is not in the document]
Comment (jeremylt) [reply interface/ceed-operator.c]: I added an example in this description.

File: backends/ref/ceed-ref-operator.c (hunk 8 above)
Comment (jedbrown) [reply backends/ref/ceed-ref-operator.c]: I guess it isn't a big deal at this point, just cruft in our model. Having some short scratch vectors into which you place the arrays might also work (if I'm context-switching back well enough that this isn't nonsense).

**Comment by jedbrown** [comment]
Seems fine to merge whenever you're ready. It'd be worth checking (or at least convincing yourself) that tests pass on GPU configurations (by skipping unsupported tests).

**Comment by jeremylt** [comment]
There is some bizarre unrelated error with `/gpu/cuda/gen`. I am investigating.

**Pull request merged**